index.bs

```diff
@@ -52,10 +66,169 @@ dictionary StorageBucketOptions {
+
+    1. If |r| is failure, then [=reject=] |p| with a {{TypeError}} and abort these steps.
+
+    1. [=/Resolve=] |p| with |r|.
+
+1. Return |p|.
+
+</div>
+
+<div algorithm>
+
+To <dfn>open a bucket</dfn> for a |storageKey| given a bucket |name| and optional |options|, run the following steps:
+
+1. Let |expires| be |options|["{{StorageBucketOptions/expires}}"] if it exists, otherwise undefined.
+
+1. If |expires| is not undefined, and is less than or equal to now, then return failure.
+
+1. Let |quota| be |options|["{{StorageBucketOptions/quota}}"] if it exists, otherwise undefined.
+
+1. If |quota| is less than or equal to zero, then return failure.
+
+1. Let |persisted| be false.
+
+1. If |options|["{{StorageBucketOptions/persisted}}"] is true, then:
+
+   1. Let |permission| be the result of [=/requesting permission to use=] "<code>persistent-storage</code>".
+
+   1. If |permission| is "{{PermissionState/granted}}", then set |persisted| to true.
+
+1. Let |bucket| be the [=/storage bucket=] named |name| in |storageKey| or null otherwise.
+
+1. If |bucket| is non-null and |bucket|'s [=bucket expiration=] is less than or equal to now, then:
+
+   1. Remove |bucket|.
+
+   1. Set |bucket| to null.
+
+1. If |bucket| is null, then:
+
+   1. Let |bucket| be a new [=/storage bucket=] with name |name|
+
+   1. Set |bucket|'s [=bucket durability|durability=] to |options|["{{StorageBucketOptions/durability}}"] if it exists.
+
+   1. Set |bucket|'s [=bucket quota|quota=] to |quota|.
+
+1. If |persisted| is true, set |bucket|'s [=/bucket mode=] to "<code>persistent</code>".
+
+1. Set |bucket|'s [=bucket expiration|expiration=] to |expires|.
+
+1. Let |storageBucket| be a new {{StorageBucket}}.
+
+1. Set |storageBucket|'s [=/storage bucket=] to |bucket|.
+
+1. Return |storageBucket|.
+
+</div>
+
+<div algorithm>
+
+To <dfn>validate a bucket name</dfn> given string |name|, run the following steps:
+
+1. If |name| contains any character that is not [=ASCII lower alpha=], [=ASCII digit=], U+005F (_), or U+002D(-), then return failure.
+
+1. If |name| [=string/length=] is 0 or exceeds 64, then return failure.
+
+1. If |name| begins with U+005F (_) or U+002D(-), then return failure.
+
+1. Return.
+
+</div>
+
+When not null, {{StorageBucketOptions/durability}} is a hint to the user agent
+specifying the desired {{StorageBucket/durability}}. The user agent MAY
+create a new {{StorageBucket}} with this [=bucket durability=]. The user agent
+MUST NOT modify the [=bucket durability=] of an existing bucket.
+
 <h3 id="storage-bucket-delete">Deleting a bucket</h3>
+
+<div algorithm>
+
+The <dfn method for="StorageBucketManager">delete(|name|)</dfn> method steps are:
+
+1. Let |environment| be [=/this=]'s [=/relevant settings object=].
+
+1. Let |storageKey| be the result of running [=obtain a storage key=] given |environment|.
+
+1. If |storageKey| is failure, then [=exception/throw=] a "{{SecurityError}}" {{DOMException}} and abort these steps.
```

Comment:
Again, `{{TypeError}}` might be preferred these days instead of these particular DOMExceptions.